Source/TCPConnection.swift

```diff
@@ -93,35 +115,71 @@ public final class TCPClientSocket: TCPSocket {
         return Data(data.prefix(bytesProcessed))
     }
 
-    public func receive(length length: Int, untilDelimiter delimiter: String, deadline: Deadline = never) throws -> Data {
+    public func receive(upTo byteCount: Int, untilDelimiter delimiter: String, deadline: Deadline = never) throws -> Data {
+        let socket = try getSocket()
         try assertNotClosed()
 
-        var data = Data.bufferWithSize(length)
+        
+        var data = Data.bufferWithSize(byteCount)
         let bytesProcessed = data.withUnsafeMutableBufferPointer {
             tcprecvuntil(socket, $0.baseAddress, $0.count, delimiter, delimiter.utf8.count, deadline)
         }
 
         try TCPError.assertNoReceiveErrorWithData(data, bytesProcessed: bytesProcessed)
         return Data(data.prefix(bytesProcessed))
     }
-
-    public func attach(fileDescriptor: FileDescriptor) throws {
-        try super.attach(fileDescriptor, isServer: false)
+    
+    public func close() -> Bool {
+        guard let socket = self.socket else {
+            closed = true
+            return true
+        }
+        
+        if closed {
+            return false
+        }
+        
+        closed = true
+        tcpclose(socket)
+        return true
+    }
+    
+    func getSocket() throws -> tcpsock {
+        guard let socket = self.socket else {
+            throw TCPError.closedSocket(description: "Socket has not been initialized. You must first connect to the socket.")
+        }
+        return socket
     }
+    
+    func assertNotClosed() throws {
```

Comment:
I don't really like the naming of this method. Maybe something like `ensureIsOpen` or `ensureNotClosed`?